Framework/Algorithms/src/SolidAngle.cpp

```diff
@@ -24,8 +26,121 @@ namespace Algorithms {
 // Register with the algorithm factory
 DECLARE_ALGORITHM(SolidAngle)
 
+namespace SolidAngleMethods {
+static const std::string GENERIC_SHAPE = "GenericShape";
+static const std::string RECTANGLE = "Rectangle";
+static const std::string VERTICAL_TUBE = "VerticalTube";
+static const std::string HORIZONTAL_TUBE = "HorizontalTube";
+static const std::string VERTICAL_WING = "VerticalWing";
+static const std::string HORIZONTAL_WING = "HorizontalWing";
+} // namespace SolidAngleMethods
+
 using namespace Kernel;
 using namespace API;
+using namespace Geometry;
+using namespace SolidAngleMethods;
+
+namespace SolidAngleHelpers {
+
+constexpr double MM_TO_METERS = 1. / 1000.;
+
+/**
+ * Returns the angle between the sample-to-pixel vector and its
+ * projection on the X-Z (vertical tube) or Y-Z (horizontal tube) plane.
+ * Note, in all cases Y is assumed to be the pointing-up direction, Z is the
+ * beam direction.
+ */
+struct AlphaAngleCalculator {
+  AlphaAngleCalculator(const DetectorInfo &detectorInfo)
+      : m_detectorInfo(detectorInfo),
+        m_samplePos(detectorInfo.samplePosition()) {}
+  double getAlpha(size_t index) const {
+    const auto sampleDetVec = m_detectorInfo.position(index) - m_samplePos;
+    auto inPlane = sampleDetVec;
+    project(inPlane);
+    return sampleDetVec.angle(inPlane);
+  }
+  virtual void project(V3D &v) const = 0;
+  virtual ~AlphaAngleCalculator() = default;
+
+private:
+  const DetectorInfo &m_detectorInfo;
+  const V3D m_samplePos;
+};
+
+struct AlphaAngleVertical : public AlphaAngleCalculator {
+  using AlphaAngleCalculator::AlphaAngleCalculator;
+  void project(V3D &v) const override { v.setY(0.0); }
```

Comment:
Using virtual functions is equivalent to (possibly worse than) the if/else branch we had before. In addition you're creating an object in the base class that is only used by 2 of the derived classes.

The ideal would be to use a template parameter in `Tube` and `Wing`. the provides the function for horizontal or vertical. Otherwise we may want to go back to the old boolean check.

Reply:
I am not convinced about your first statement. Previously we had if/else in the function which is called for each pixel. Now once the calculator is of right sub-type, I would expect override resolution be faster. Of course we can measure this, but if it is not worse, I think now it is the most elegant.

Reply:
The vtable is a branch, and *may* be harder for the compiler to inline. 

At this point getting the tests passing and the functionality is much more important than trying to save a few hundredths of a second.